src/node/node-utils.ts

```diff
@@ -80,34 +93,116 @@ export function updateMoveToml(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let lockJson: any;
+
+    // Version < 1 (non-upgrade)
+    if (!version) {
+        // Update the published-at field under the package section e.g. published-at = "0x01"
+        // (only applied to non-upgraded versions)
+        (tomlJson as Record<string, Record<string, string>>).package['published-at'] = packageId;
+
+        // Update the package address under the addresses section e.g. gas_service = "0x1"
+        (tomlJson as Record<string, Record<string, string>>).addresses[packageName] = packageId;
+    } else {
+        if (!wasBuilt) {
+            throw new Error(`Upgrade parameter missing, no lock file was found for given path: ${lockPath}`);
+        }
 
-    // Update the package address under the addresses section e.g. gas_service = "0x1"
-    (moveJson as Record<string, Record<string, string>>).addresses[packageName] = packageId;
+        // Version >= 1
+        // Remove the 'published-at' field (if required)
+        let legacyPackageId;
+
+        if (originalPackageId) {
+            legacyPackageId = originalPackageId;
+        } else if ((tomlJson as Record<string, Record<string, string>>).package['published-at']) {
+            legacyPackageId = (tomlJson as Record<string, Record<string, string>>).package['published-at'];
+            if (!legacyPackageId) throw new Error(`Upgrade parameter missing, no published-at id was found for given path: ${tomlPath}`);
+            delete (tomlJson as Record<string, Record<string, string>>).package['published-at'];
+        }
+
+        // Reset the package address in the addresses field to "0x0"
+        (tomlJson as Record<string, Record<string, string>>).addresses[packageName] = emptyPackageId;
+
+        // Update the lock file (lock file must exist)
+        // Read the Move.lock file
+        const lockRaw = fs.readFileSync(lockPath, 'utf8');
+        // Parse the Move.lock file as JSON
+        lockJson = toml.parse(lockRaw);
+
+        // If syncing an already published upgrade, use 'original-published-id' in place of 'published-at'
+        // for deriving the legacy package id
+        if (!legacyPackageId) {
+            try {
+                legacyPackageId = (lockJson as Record<string, Record<string, string>>)[`env.${network}`]['original-published-id'];
+            } catch {
+                throw new Error(`Upgrade parameter missing, no original-published-id was found for given path: ${lockPath}`);
+            }
+        }
+
+        // Add the required sections for building versioned dependencies
+        // [env]
+        if (!lockJson.env) {
+            lockJson.env = {};
+        }
+
+        if (network !== 'devnet' && network !== 'testnet' && network !== 'mainnet') {
+            throw new Error(
+                `Unrecognized chain-id for contract upgrade for given network ${network}. Must be one of ${JSON.stringify(chainIds)}`,
+            );
+        }
```

Comment:
do such validation as early as possible, in this case near the top of the function